File: Leetcode/104.MaximumDepthofBinaryTree/104/104.c
/* ************************************************************************
> File Name:     104.c
> Author:        Jack Yang
> Created Time:  2023年05月05日 (週五) 21時38分36秒
> Description:   Maximum Depth of Binary Tree 
 ************************************************************************/
#include<stdio.h>
#include<stdlib.h>
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};
typedef struct TreeNode TreeNode;

void sol(TreeNode *ptr,int level,int *result){
    if(ptr==NULL){
        if(level>*result){
            *result=level;
        }
        return;
    }
    sol(ptr->left,level+1,result);
    sol(ptr->right,level+1,result);
    return;
}

int maxDepth(struct TreeNode* root){
    if(root==NULL)return 0;
    int level=0;
    int result=0;
    sol(root,level,&result);
    return result;
}

TreeNode *NewNode(int x){
    TreeNode *node=(TreeNode *)malloc(sizeof(TreeNode));
    node->val=x;
    node->left=NULL;
    node->right=NULL;
}
int main(int argc,char *argv[]){
    TreeNode *root=NewNode(3); 
    TreeNode *ptr=root;
    ptr->left=NewNode(9);
    ptr->right=NewNode(20);
    ptr=ptr->right;
    ptr->left=NewNode(15);
    ptr->right=NewNode(7);
    int result=maxDepth(root);
    printf("%d",result);
    return 0;
}

File: Leetcode/104.MaximumDepthofBinaryTree/104/104.py
from typing import Optional
import queue
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        que=queue.Queue()
        que.put(root)
        result=0
        while not que.empty():
            size=que.qsize()
            result+=1
            for i in range(size):
                node=que.get()
                if node.left:
                    que.put(node.left)
                if node.right:
                    que.put(node.right)
        return result
root=TreeNode(3)
root.left=TreeNode(9)
root.right=TreeNode(20)
ptr=root.right
ptr.left=TreeNode(15)
ptr.right=TreeNode(7)
sol=Solution()
print(sol.maxDepth(root))
